### PR TITLE
Unblock synchronous signals when in use and ensure registering handler before unblocking a signal 

### PR DESCRIPTION
**Pull request by rmnattas**
When signals are blocked, it stays blocked across sub-processes through inheritance of the SignalMask.
Keeping synchronous signals blocked cause external effect with specific environments on signal control causing unexpected change in signal mask/handler [1].

Allowing synchronous signals to get unblocked when used, as asynchronous signals do, fix the issue and ensure signals don't get received by a wrong handler and for the process to terminates abruptly. This PR also moves unblocking signals to after registering the signal handler with the OS instead of before to ensure no signals get received between the unblocking and registering of signals.

> As per the Notes in https://man7.org/linux/man-pages/man2/sigprocmask.2.html, there are only side-effects for blocking signals such as SIGBUS, SIGFPE, SIGILL and SIGSEGV . Unblocking signals doesn't have any side-effects. So, we can unblock all signals (synchronous + asynchronous) after registering a handler against the signal in OMR. This can be treated as a permanent fix. This will also add consistency to unblocking signals. [2]

[1]
https://github.com/eclipse/openj9/issues/10744#issuecomment-750893654
[2]
https://github.com/eclipse/openj9/issues/10744#issuecomment-754907388

Issue: eclipse/openj9#10744
Signed-off-by: Abdulrahman Alattas <rmnattas@gmail.com>

### CODE REVIEW

**Comment by rmnattas** [comment]
fyi @babsingh @gita-omr

**Comment by babsingh** [comment]
@rmnattas there is a test failure:

```
2021-01-07T14:57:46.3894460Z 31: [----------] 1 test from PortSignalExtendedTests
2021-01-07T14:57:46.3994930Z 31: 	/Users/runner/work/1/s/fvtest/porttest/omrsignalExtendedTest.cpp line  536: omrsig_ext_test1 	masked thread received signal as not expected.(SignalStatus=2)
2021-01-07T14:57:46.4096380Z 31: 
2021-01-07T14:57:46.4197530Z 31: 			LastErrorNumber: -108
2021-01-07T14:57:46.4298210Z 31: 			LastErrorMessage: No such file or directory
2021-01-07T14:57:46.4399170Z 31: 
2021-01-07T14:57:46.4500310Z 31: 	omrsig_ext_test1	:test stopped with errors
```

**Comment by babsingh** [comment]
Copyright needs to be updated.

**Comment by rmnattas** [comment]
@babsingh The failing test sets a mask to block SIGILL, creates testing threads which inherence that mask and perform few tests, two of which relevant.

1. Checking the signal mask and ensuring SIGILL is blocked, which passes.
2. Installing a handler for SIGILL and send a SIGILL signal to that thread and expecting it to not be received [1], which fails with this PR.

I'm not fully sure of what the root of this test wanted to check by this behaviour, but its not compatible with this PR.
Given that the issue is an external effect and unblocking the signal is a way of stopping it from happening, its either finding another way of eliminating the external effect (i.e. maybe immediately unblock then block the synchronous signals) or change this test.

[1]
https://github.com/eclipse/omr/blob/d2a0060eca78f8e21054b7d2f064d5769af8909b/fvtest/porttest/omrsignalExtendedTest.cpp#L505-L539

**Comment by babsingh** [comment]
@rmnattas Can we change the order of setting the signal mask and installing the handler in the test? It will have the same impact. The test only wants to verify the impact of calling `pthread_sigmask` from a thread. It does not outline the order of installing the handler and calling `pthread_sigmask`. Until `pthread_sigmask` is invoked at some point, the test should stay valid. My vote will be to change the test. Do you agree?

**Test description:**
https://github.com/eclipse/omr/blob/d2a0060eca78f8e21054b7d2f064d5769af8909b/fvtest/porttest/omrsignalExtendedTest.cpp#L330-L335

**Comment by rmnattas** [comment]
@babsingh Yes, I think thats good idea. I agree especially because there's no other apparent fix given that the issue is an external effect.

**Comment by babsingh** [comment]
@genie-omr build all

**Review by babsingh — APPROVED** [review]
For the test, you need to mention the reason for changing `SIGILL` to `SIGSEGV` in the amended test case. You did specify the reason while chatting with me. You should include it in the commit message.

These changes LGTM. Tentatively approving this PR on account that the documentation of the test case will be improved.

For some reason, the PR builds didn't launch on my command. Delegating a committer (@0xdaryl) for launching PR builds once the documentation for the test case is updated.

@tajila We are unblocking synchronous signals (SIGSEGV, ILL, FPE, BUS, etc.) after registering a handler for them in `omrsig_protect`. C/C++ applications may block the sync signal and launch the JVM. With these changes, the JVM will undo their block if the JVM registers a handler via `omrsig_protect` for the blocked synchronous signal. There is no spec requirement which prevents us to do this change. Also, I feel that native apps will rarely block synchronous signals, and instead, they should use `-Xrs` to disable handlers. Do you notice any issues with this change?

**Comment by 0xdaryl** [comment]
@genie-omr build all

**Comment by rmnattas** [comment]
Failure cross compiling the arm build

```
10:15:33  checking build system type... x86_64-pc-linux-gnu
10:15:33  checking host system type... x86_64-pc-linux-gnu
10:15:33  checking for gcc... arm-linux-gnueabihf-gcc
10:15:33  checking whether the C compiler works... no
10:15:33  configure: error: in `/home/jenkins/workspace/Build':
10:15:33  configure: error: C compiler cannot create executables
10:15:33  See `config.log' for more details
10:15:33  run_configure.mk:164: recipe for target 'CONFIGURE_SENTINEL_FILE' failed
10:15:33  make: *** [CONFIGURE_SENTINEL_FILE] Error 77
```

**Comment by tajila** [comment]
>  Do you notice any issues with this change?

Nothing stands out to me ATM. I'm a little worried about the impact of this change to users who run the the JVM in isolated environment using custom native launchers. However, I'm not aware of any user blocking native signals. Since the spec doesn't prohibit this change then users can not rely on any consistent behaviour here. 

Do you know what the behaviour of the RI is?

**Comment by rmnattas** [comment]
### Behaviour testing
Using a native C app which **sets a handler for and blocks** all synchronous signals* then it execute Java code and sleep, during which a signal gets send to the native app (test modified from https://github.com/eclipse/openj9/issues/8735#issue-574912507)
We see the following behaviour (all synchronous signals* had the same behaviour):

|                 | All Synchronous Signals* |
|-----------------|--------------------------|
| **Default Setting** |                          |
| RI              | Received by app handler  |
| OpenJ9 Before Change        | Nothing (blocked)        |
| OpenJ9 After Change           | JVM core dump            |
|                 |                          |
| **With `-Xrs`**       |                          |
| RI              | Received by app handler  |
| OpenJ9 Before Change          | Nothing (blocked)        |
| OpenJ9 After Change           | Nothing (blocked)        |

RI apparently unblocks these signals but it doesn’t have the JVM handling them.
OpenJ9 before this change and with `-Xrs` kept the signals blocked, and after this change it unblocked them but they got handled by the JVM.

\* `SIGSEGV` `SIGBUS` `SIGILL` `SIGFPE` `SIGTRAP`

**Comment by tajila** [comment]
Okay, it looks like this change puts us more in line with the RI when using default options. I'd say go ahead with this.

**Comment by babsingh** [comment]
@0xdaryl This change is good to be merged.

**Review by 0xdaryl — APPROVED** [review]

**Comment by gita-omr** [comment]
@zl-wang fyi

**Pull request merged**
